packages/eas-cli/src/commandUtils/EasCommand.ts

```diff
@@ -12,26 +14,75 @@ import {
   initAsync as initAnalyticsAsync,
   logEvent,
 } from '../analytics/rudderstackClient';
-import { learnMore } from '../log';
+import Log, { learnMore } from '../log';
+import { interactiveOra, nonInteractiveOra } from '../ora';
 import { findProjectRootAsync } from '../project/projectUtils';
+import { InteractivePrompts, NonInteractivePrompts, Prompts } from '../prompts';
 import { getUserAsync } from '../user/User';
 import { ensureLoggedInAsync } from '../user/actions';
 import { easCliVersion } from '../utils/easCli';
 import { setVcsClient } from '../vcs';
 import GitClient from '../vcs/clients/git';
 
+type CommandOptions =
+  | {
+      nonInteractive: true;
+      json: true;
+    }
+  | {
+      nonInteractive: boolean;
+      json: false;
+    };
+
+export type CommandContext = Readonly<
+  CommandOptions & {
+    logger: Log;
+    prompts: Prompts;
+    ora: (options?: Options | string) => Ora;
+  }
+>;
+
 export default abstract class EasCommand extends Command {
+  static override globalFlags = {
+    json: Flags.boolean({
+      description: 'Enable JSON output, non-JSON messages will be printed to stderr.',
+      dependsOn: ['non-interactive'],
+    }),
+    'non-interactive': Flags.boolean({
+      description: 'Run the command in non-interactive mode.',
+    }),
+  };
+
+  // disable built in oclif json stuff
+  enableJsonFlag = false;
+
   /**
    * When user data is unavailable locally, determines if the command will
    * force the user to log in
    */
   protected requiresAuthentication = true;
   protected mustBeRunInsideProject = true;
 
-  protected abstract runAsync(): Promise<any>;
+  protected abstract runAsync({
+    nonInteractive,
+    json,
+    logger,
+    prompts: {
+      promptAsync,
+      confirmAsync,
+      selectAsync,
+      toggleConfirmAsync,
+      pressAnyKeyToContinueAsync,
+    },
+    ora,
```

Comment:
I really like context objects. However, I'm afraid that because of how we use prompts/logger/ora in our codebase, it'll be challenging to use those three everywhere. I see that you updated two commands (`branch:delete` and `device:delete`) to use the proposed solution. I think we should rather start by seeing how difficult/time-consuming the refactor would be for our two most complicated commands - `build` and `submit`. I'm especially interested in seeing how we could use the new context object in the credentials code.

An alternative approach that is not so elegant as the one proposed by you but probably doesn't require a huge refactor and isn't that bad after all:

Instead of passing the context with `logger`/`prompts`/`ora` to every command. We could have a global state in each of the modules. The global state would be set in `EasCommand`.

An example of `log.ts` would be:
```
// default to interactive class
let Log = InteractiveLog;

function setInteractivity(interactive: boolean) {
  if (interactive) { 
    Log = InteractiveLog;
  } else {
    Log = NonInteractiveLog;
  }
}
```

and later in `EasCommand`:
```
  // ...
  setLogInteractivity(!nonInteractive);
  setOraInteractivity(!nonInteractive);
  setPromptsInteractivity(!nonInteractive);
  // ...
```

Let me know what you think.

Reply:
I usually am opposed to global state, but due to the nature of this program (memory wiped each execution) I'm less opposed.